app/decorators/user_decorator/role.rb

```diff
@@ -2,7 +2,9 @@
 
 module UserDecorator
   module Role
-    def roles
+    FIRSTWEEK = 7
+
+    def roles(user = current_user)
```

Comment:
@harada-webdev 
質問なんですが、

```suggestion
    def roles
```

とした上で、

```ruby
roles.unshift('new-user') if new_student_or_trainee && current_user&.mentor?
```

にすると都合が悪くなるケースってありますか？

Reply:
@JunichiIto 
この場合だと、`roles`を使用している既存のテストで、`current_user`が未定義となりエラーが発生します。
しかし、これだとunikounioさんの仰る通り、テスト側の事情を汲むために、本番の実装に特別な処理が入っている状態になっているので、テスト側で`current_user`を定義する方法を考えているのですが、`stub`などを使って`current_user`を定義しても、未定義のエラーが起こっている現状で悩んでいます。

Reply:
@JunichiIto 
この場合ですと、`roles`を使用する一部の既存のテストで、`current_user`が未定義となりエラーが発生します。

具体的には、テスト環境から、`app/views/activity_mailer`以下と`app/components`以下にある`roles`や`primary_role`などを実行しようとすると、それらのメソッドで必要になる`current_user`を参照できず未定義となりエラーが発生します。また、`stub`を使ってテスト内で`current_user`を定義しても、同様にエラーとなります。
(一方で開発環境では、`app/components`以下からの場合には、`current_user`を参照することができます。)

そのため、上記のようなテストから`current_user`を参照させないために、`roles`の引数を`(user = current_user)`のようにして、上記のテストから`roles`などを実行する際には、レシーバを引数にとっている現状です。

Reply:
回答ありがとうございます。

うーむ、なるほど・・・と思って、とりあえず手元でテストを回して状況を確認してみようとしたのですが、なぜか僕のローカルでは以下のようにsystem testがテストサーバーにアクセスできず、再現を断念しました😭
（こんなエラー見たことない）
![Screenshot 2025-03-02 at 8 50 21](https://github.com/user-attachments/assets/5c7f7b54-2ac8-41a3-9b23-8de216cae303)
実行したコマンドはこれ

```
HEADFUL=1 PARALLEL_WORKERS=1 rails test test/system/users_test.rb:695
```

----

というわけで、あまり技術的な詳細を把握しないまま、一般論としてコメントします。

まず、すでにあちこちで使われているメソッドに引数を増やし、既存の呼び出し元にぷちぷちと引数を足していく、というのは、追加漏れが発生しやすいのでなるべく避けたいです。

また引数を増やすことは複雑度を上げることにもなるので、将来の保守性や可読性を悪化させます。

つまり、ひとことで言うと、この対応は費用対効果が悪いです。

----

では、どうすべきかというと、プログラム側で頑張るのではなく、仕様を簡単にする、というアプローチを検討してみましょう。

今回であればおそらく「メンターにしか表示されない」という部分が厄介なんだと思います。
しかし、「入ってまだ1週間」という情報はそんなにクリティカルな情報でしょうか？
個人的な意見としては、「この人は新人さんですよ」と表示するだけなので、これがメンター以外の人に見せてはいけない情報だとは思いません。

おそらくですが、駒形さんから最初に出た要望が「入ったばかりのユーザーなのでサポートやコメントを多めにお願いします」と画面に表示したい、という内容だったので、それに引きずられているだけだと思います。
このコメントであればたしかにメンターにだけ表示すべきですが、途中で要件が変わった（＝アイコン表示を変えるだけになった）ので、「メンターにしか表示されない」の部分もあわせて再検討すべきです。

「メンターにしか表示されない」を簡単に実現できる場合は別ですが、そうでないなら、「入って1週間のユーザーには（メンターであるかどうかを問わず）何らかのマークを表示する」とした方が、費用対効果が高いと思います。

Reply:
@JunichiIto 
ご提案ありがとうございます！
確かにいろいろな場所での引数の追加によって保守性や可読性が低くなってしまうのでプログラム側ではなく仕様を変えたほうが良いですね
また、新入生アイコンの表記が、メンター以外にもわかるようになれば、新入生の日報やコメントなどが皆から注目されやすくなる利点があるとも思います
ですので、この件について、駒形さんと町田さんにも確認を行うことにします

Reply:
@harada-webdev 

ユーザーのroleが見ている人によって変わるのはおかしいと思います。
（new-userは誰から見てもnew-user）

ここを変えるのではなく、 @machida さんと連携してCSSと絡めて実現するのがいいと思います。
（メンターが見ている場合、`.new-user`の画像は「赤い縁をつける」など）

Reply:
@komagata 
方針が先週の質問タイムでお話したことと少し変わったため、確認のため連絡いたします。(町田さんからは既に了承済みです)

現在の方針:
``` slim
span class="a-user-role is-#{user.primary_role} is-#{joining_status}"
```
上記のように`a-user-role` が含まれている全てのclassに`is-#{joining_status}`を追加して`is-new-user`を反映するという形なのですが、これだとすべてのユーザーアイコンに対して`is-new-user`が反映されているかのテストを行う必要があり、テストの量が膨大になってしまうため方針を変更いたしました。

変更後:
```slim
span class="a-user-role is-#{user.primary_role}"
```

見ているユーザーが入って一週間以内の現役生と研修生の場合、`user.primary_role`は`is-new-student`と`is-new-trainee`を返します。そして、「メンターにしか表示されない」という条件については、メンターモードで使用している`bodyタグ`の`is-mentor`classを使用してCSS側で判断します。変更後の方針ですと、`user.primary_role`が`is-new-student`と`is-new-trainee`を返すかどうかのテストで済ますことが可能となります。

こちらの方針でよろしいでしょうか？

Reply:
@harada-webdev 

> これだとすべてのユーザーアイコンに対してis-new-userが反映されているかのテストを行う必要があり、テストの量が膨大になってしまうため方針を変更いたしました。

ちょっと**膨大**がどのくらいの数を示しているのかわからないですがunit testであればテストがたくさんになっても問題ないと思います。

> 見ているユーザーが入って一週間以内の現役生と研修生の場合、user.primary_roleはnew-studentとnew-traineeを返します。そして、「メンターにしか表示されない」という条件については、メンターモードで使用しているbodyタグのis-mentorclassを使用してCSS側で判断します。変更後の方針ですと、user.primary_roleがis-new-studentとis-new-traineeを返すかどうかのテストで済ますことが可能となります。

意味としてnew-userかold-userかというのはroleではないように思います。

Reply:
@komagata 
```slim
span class="a-user-role is-#{user.primary_role} is-#{joining_status}"
```
例えば、ユーザーが現役生の場合に、上記のクラスが`a-user-role is-student is-new-user`と正しく表示されるかどうかをユニットテストで行い

日報や提出物などの特定のページでのアイコンがnew-userのマークになっているかの証拠は、自動のsystemテストではなく、スクショや動画など手動のテストでそれぞれ確認していけばよいということですか？